source/adios2/engine/hdf5/HDF5WriterP.cpp

```diff
@@ -269,71 +222,100 @@ void HDF5Writer::Write(const std::string variableName,
+void HDF5Writer::Advance(float timeout_sec) { _H5File.H5_Advance(0); }
+
 void HDF5Writer::Close(const int transportIndex)
 {
-    // std::cout << " ===> CLOSING HDF5 <===== " << std::endl;
-    // H5Dclose(_dset_id);
-    // H5Sclose(_filespace);
-    // H5Sclose(_memspace);
-    // H5Pclose(_plist_id);
-    H5Fclose(_file_id);
+    /*
+    //void* hi = H5Iobject_verify(H5S_SCALAR, H5I_DATASPACE);
+
+    hid_t s = H5Screate(H5S_SCALAR);
+    //hid_t attr = H5Acreate(_H5File.m_Group_id, "NumTimeSteps",
+    H5T_NATIVE_UINT,
+    s, H5P_DEFAULT, H5P_DEFAULT);
+    hid_t attr = H5Acreate(_H5File.m_File_id, "NumTimeSteps", H5T_NATIVE_UINT,
+    s,
+    H5P_DEFAULT, H5P_DEFAULT);
+    uint  totalts = _H5File.m_CurrentTimeStep+1;
+    H5Awrite(attr,H5T_NATIVE_UINT,&totalts);
+
+    H5Sclose(s);
+    H5Aclose(attr);
+    */
+    _H5File.H5_Close();
 }
 
 template <class T>
 void HDF5Writer::UseHDFWrite(Variable<T> &variable, const T *values,
                              hid_t h5type)
 {
+    _H5File.CheckWriteGroup();
     // here comes your magic at Writing now variable.m_UserValues has the data
     // passed by the user
     // set variable
     variable.m_AppValues = values;
     m_WrittenVariables.insert(variable.m_Name);
 
-    int dimSize = variable.m_GlobalDimensions.size();
-    /*
-    std::cout << "writting : " << variable.m_Name
-              << " dim size:" << variable.m_GlobalDimensions.size() <<
-    std::endl;
-
-    for (int i = 0; i < dimSize; i++) {
-      std::cout << " dim: " << i << ", size:" << variable.m_GlobalDimensions[i]
-                << " offset=" << variable.m_Offsets[i]
-                << " count=" << variable.m_LocalDimensions[i] << std::endl;
-    }
-    */
+    int dimSize = std::max(variable.m_GlobalDimensions.size(),
+                           variable.m_LocalDimensions.size());
+
     std::vector<hsize_t> dimsf, count, offset;
 
     for (int i = 0; i < dimSize; i++)
     {
-        dimsf.push_back(variable.m_GlobalDimensions[i]);
-        count.push_back(variable.m_LocalDimensions[i]);
-        offset.push_back(variable.m_Offsets[i]);
+        if (variable.m_GlobalDimensions.size() == dimSize)
+        {
+            dimsf.push_back(variable.m_GlobalDimensions[i]);
+        }
+        else
+        {
+            dimsf.push_back(variable.m_LocalDimensions[i]);
+        }
+
+        if (variable.m_LocalDimensions.size() == dimSize)
+        {
+            count.push_back(variable.m_LocalDimensions[i]);
+            if (variable.m_Offsets.size() == dimSize)
+            {
+                offset.push_back(variable.m_Offsets[i]);
+            }
+            else
+            {
+                offset.push_back(0);
+            }
+        }
+        else
+        {
+            count.push_back(variable.m_GlobalDimensions[i]);
+            offset.push_back(0);
+        }
     }
 
-    _filespace = H5Screate_simple(dimSize, dimsf.data(), NULL);
+    hid_t _filespace = H5Screate_simple(dimSize, dimsf.data(), NULL);
```

Comment:
See https://github.com/ornladios/ADIOS2/wiki/Variables-Scope,-Functions,-and-Namespaces for local variables using `camelCase`